src/Algebra/Properties/Semigroup.agda

```diff
@@ -0,0 +1,122 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some theory for Semigroup
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra using (Semigroup)
+open import Algebra.FunctionProperties using (Commutative)
+import      Algebra.Properties.Setoid as OfSetoid
+open import Function using (_$_)
+import      Relation.Binary.EqReasoning as EqR
+
+
+--****************************************************************************
+module Algebra.Properties.Semigroup {ℓ ℓ=} (H : Semigroup ℓ ℓ=) where
+
+open Semigroup H using (_≈_; _∙_; assoc; ∙-congˡ; ∙-congʳ; setoid)
+open OfSetoid setoid public
+
+open EqR setoid
+
+∙-assoc = assoc   -- to export
+
+
+--============================================================================
+module _ (comm : Commutative _≈_ _∙_)
```

Comment:
1) For Bin0, only one or two of them are needed. For Bin/*, more of them
   are needed, for BFLib with Fraction almost all of them are needed.

2) Further modules will require to insert into  Properties/Semigroup
   several lemmata as ``∣trans :  Transitive _∣_``,
   which are proved not only for the commutative case.
   And standard library does not need to require commutativity for the proofs
   which do not rely on commutativity.

Remain with the Semigroup head parameter?

Reply:
The file `Algebra.Properties.Semigroup` is fine. It's just these proofs which are problematic as they should belong in `Algebra.Properties.CommutativeSemigroup`.

Reply:
I am going to put this into something like this:
```
module ... Algebra.Properties.CommutativeSemigroup (H : Semigroup _ _)
                                                (open Semigroup H using (_ ≈_;  _∙ _))                                                                                                                                                                        
                                                (comm : Commutative _ ≈_  _∙ _)  where  
```
All right?
Introducing ``record _ CommutativeSemigroup`` solely to add a single field does not look nice, because there are already many levels in the algebraic record hierarchy.

Reply:
That's fine for now. We can always add CommutativeSemigroup later.